aws/rust-runtime/aws-config/src/profile/parser.rs

```diff
@@ -143,35 +167,151 @@ impl ProfileSet {
-///
-/// An AWS config may be composed of a multiple named profiles within a [`ProfileSet`].
+/// Represents a top-level section (e.g., `[profile name]`) in a config file.
+pub(crate) trait Section {
+    /// The name of this section
+    fn name(&self) -> &str;
+
+    /// Returns all the properties in this section
+    fn properties(&self) -> &HashMap<String, Property>;
+
+    /// Returns a reference to the property named `name`
+    fn get(&self, name: &str) -> Option<&str>;
+
+    /// True if there are no properties in this section.
+    fn is_empty(&self) -> bool;
+
+    /// Insert a property into a section
+    fn insert(&mut self, name: String, value: Property);
+}
+
 #[derive(Debug, Clone, Eq, PartialEq)]
-pub struct Profile {
+struct SectionInner {
     name: String,
     properties: HashMap<String, Property>,
 }
 
+impl Section for SectionInner {
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn properties(&self) -> &HashMap<String, Property> {
+        &self.properties
+    }
+
+    fn get(&self, name: &str) -> Option<&str> {
+        self.properties
+            .get(to_ascii_lowercase(name).as_ref())
+            .map(|prop| prop.value())
+    }
+
+    fn is_empty(&self) -> bool {
+        self.properties.is_empty()
+    }
+
+    fn insert(&mut self, name: String, value: Property) {
+        self.properties
+            .insert(to_ascii_lowercase(&name).into(), value);
+    }
+}
+
+/// An individual configuration profile
+///
+/// An AWS config may be composed of a multiple named profiles within a [`ProfileSet`].
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Profile(SectionInner);
+
 impl Profile {
     /// Create a new profile
-    pub fn new(name: String, properties: HashMap<String, Property>) -> Self {
-        Self { name, properties }
+    pub fn new(name: impl Into<String>, properties: HashMap<String, Property>) -> Self {
+        Self(SectionInner {
+            name: name.into(),
+            properties,
+        })
     }
 
     /// The name of this profile
     pub fn name(&self) -> &str {
-        &self.name
+        self.0.name()
     }
 
     /// Returns a reference to the property named `name`
     pub fn get(&self, name: &str) -> Option<&str> {
-        self.properties
-            .get(to_ascii_lowercase(name).as_ref())
-            .map(|prop| prop.value())
+        self.0.get(name)
+    }
+}
+
+impl Section for Profile {
+    fn name(&self) -> &str {
+        self.0.name()
+    }
+
+    fn properties(&self) -> &HashMap<String, Property> {
+        self.0.properties()
+    }
+
+    fn get(&self, name: &str) -> Option<&str> {
+        self.0.get(name)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn insert(&mut self, name: String, value: Property) {
+        self.0.insert(name, value)
+    }
+}
+
+/// A `[sso-session name]` section in the config.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct SsoSession(SectionInner);
```

Comment:
do we want this public? One might imagine a different API like:

```rust
let x: ProfileSet = ...;
let subsection: Option<&Profile> = x.get_extra_section("sso-session);
```

This keeps our public API narrower and doesn't hard code the concept of sso-sessions into the (currently) generic profile file parser.

Internally the profile parser _could_ still store the same repr, or we could store `extra_sections: HashMap<String, Profile>`

Maybe I'm missing something about the API that makes this impossible?

On a side note, we don't actually need to make this code public _at all_ since it's only needed internally.

Reply:
What makes it an "extra session" though? My understanding is that anything that is unrecognized becomes a profile name.

Reply:
Hmmm. From what I can see in the profile parser tests, it actually seems to ignore anything with a space in it if it doesn't start with `profile`.

Reply:
The parser is already hardcoded to understand profiles, so I don't see this as that big of a deal. If we were to start over again, a generic approach would probably be better. But given we need to maintain backwards compatibility, I think this is fine for now. I made the sso-session code private so that this can be a two-way door.

Reply:
I guess I see the semantics a little different. Profiles being the fundamental unit because you're always reading the file in the "scope" of a given profile + additional global metadata accessible to all profiles.

But as you said, two way door, I'm fine with this.